src/main/kotlin/org/rust/ide/surroundWith/RustExpressionSurrounderBase.kt

```diff
@@ -0,0 +1,34 @@
+package org.rust.ide.surroundWith
+
+import com.intellij.lang.surroundWith.Surrounder
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiElement
+import org.rust.lang.core.psi.RustExprElement
+import java.util.*
+
+abstract class RustExpressionSurrounderBase : Surrounder {
+    abstract fun isApplicable(expression: RustExprElement): Boolean
+    abstract fun surroundExpression(project: Project, editor: Editor, expression: RustExprElement): TextRange?
+
+    final override fun isApplicable(elements: Array<out PsiElement>): Boolean {
+        if (elements.size != 1 || elements[0] !is RustExprElement) {
+            return false
+        }
+
+        val expression = elements[0] as RustExprElement
+
+        // TODO: Some additional filtering may be required.
+
+        return isApplicable(expression)
+    }
+
+    final override fun surroundElements(project: Project, editor: Editor, elements: Array<out PsiElement>): TextRange? {
+        require(elements.size == 1 && elements[0] is RustExprElement) {
```

Comment:
Which proves that `elements[0] as RustExprElement` on the next line is valid. This is exactly my point: if you don't separate "look" and "leap", they can't get out of sync. 

If you define

```
fun target(elements: Array<out PsiElement>): RustExprElement? =
    elements.singleOrNull() as? RustExprElement
```

you would be able to reuse it in `isApplicable` and in `surroundElements`.